File: android/src/main/java/com/cardihab/app/auth0capacitor/Auth0CapacitorPlugin.java
package com.cardihab.app.auth0capacitor;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "Auth0Capacitor")
public class Auth0CapacitorPlugin extends Plugin {

    private Auth0Capacitor implementation;

    @Override
    public void load() {
        implementation = new Auth0Capacitor(this.getActivity().getApplicationContext());
    }

    @PluginMethod
    public void createAuth0Client(PluginCall call) {
        String domain = call.getString("domain");
        String clientId = call.getString("clientId");
        String scheme = call.getString("scheme", "com.myclubsmyscores.ClaytargetApp");

        implementation.createAuth0Client(domain, clientId, scheme, call);
    }

    @PluginMethod
    public void login(PluginCall call) {
        String scope = call.getString("scope");
        String audience = call.getString("audience");

        implementation.login(scope, audience, call);
    }

    @PluginMethod
    public void handleRedirect(PluginCall call) {
        call.unimplemented("Not implemented on Android.");
    }

    @PluginMethod
    public void renew(PluginCall call) {
        call.unimplemented("Not implemented on Android.");
    }

    @PluginMethod
    public void logout(PluginCall call) {
        implementation.logout(call);
    }

    @PluginMethod
    public void getUser(PluginCall call) {
        implementation.getUser(call);
    }

    @PluginMethod
    public void getTokenSilently(PluginCall call) {
        call.unimplemented("Not implemented on Android.");

    }
}

File: ios/Plugin/Auth0CapacitorPlugin.swift
import Foundation
import Capacitor

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(Auth0CapacitorPlugin)
public class Auth0CapacitorPlugin: CAPPlugin {
    private let implementation = Auth0Capacitor()

    @objc func createAuth0Client(_ call: CAPPluginCall) {
       implementation.domain = call.getString("domain") ?? ""
       implementation.clientId = call.getString("clientId") ?? ""

        call.resolve()
    }

    @objc func login(_ call: CAPPluginCall) {
        let scope = call.getString("scope") ?? ""
        let audience = call.getString("audience") ?? ""
//        let appState = call.getString("appState") ?? ""

        implementation.login(scope, audience) { result in
            switch result {
                case .failure(let error):
                    // Handle the error
                    print("Error: \(error)")
                    call.reject(error.localizedDescription)
                case .success(let credentials):
                    call.resolve([
                        "accessToken": credentials.accessToken!,
                        "idToken": credentials.idToken!,
                        "refreshToken": credentials.refreshToken!,
                        "expiresIn": credentials.expiresIn!
                    ])
            }
        }
    }

    @objc func renew(_ call: CAPPluginCall) {
        implementation.renew { result in
            switch result {
                case .failure(let error):
                    // Handle the error
                    print("Error: \(error)")
                    call.reject(error.localizedDescription)
                case .success(let credentials):
                    call.resolve([
                        "accessToken": credentials.accessToken!,
                        "idToken": credentials.idToken!,
                        "refreshToken": credentials.refreshToken!,
                        "expiresIn": credentials.expiresIn!
                    ])

            }
        }
    }

    @objc func getAccessTokenSilently(_ call: CAPPluginCall) {
        implementation.getAccessTokenSilently { result in
            switch result {
                case .failure(let error):
                    // Handle the error
                    print("Error: \(error)")
                    call.reject(error.localizedDescription)
                case .success(let credentials):
                    call.resolve([
                        "accessToken": credentials.accessToken!,
                        "idToken": credentials.idToken!,
                        "refreshToken": credentials.refreshToken!,
                        "expiresIn": credentials.expiresIn!
                    ])

            }
        }
    }

    @objc func logout(_ call: CAPPluginCall) {
        implementation.logout()
        call.resolve()
    }

    @objc func getUser(_ call: CAPPluginCall) {
        implementation.getUser { result in
            switch result {
            case .failure(let error):
                // Handle the error
                print("Error: \(error)")
                call.reject(error.localizedDescription)
            case .success(let profile):
                call.resolve([
                    "givenName": profile.givenName,
                    "familyName": profile.familyName,
                    "birthDate": profile.birthdate,
                    "gender": profile.gender
                ])
            }
        }
        
    }

}
